File: LeetCode/133.克隆图.java
/*
 * @lc app=leetcode.cn id=133 lang=java
 *
 * [133] 克隆图
 */

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.HashMap;
import java.util.List;
import java.util.Queue;
// @lc code=start

// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return null;
        Node res = new Node(node.val);
        HashMap<Node, Node> map = new HashMap<>();
        map.put(node, res);
        Queue<Node> loc = new LinkedList<>();
        loc.add(node);
        while (!loc.isEmpty())
        {
            Node cur = loc.peek();
            loc.poll();
            for (Node neighbor: cur.neighbors)
            {
                if (!map.containsKey(neighbor))
                {
                    Node clone = new Node(neighbor.val);
                    map.put(neighbor, clone);
                    loc.add(neighbor);
                }
                map.get(cur).neighbors.add(map.get(neighbor));
            }
        }
        return res;
    }
}
// @lc code=end

File: LeetCode/133.克隆图.cpp
/*
 * @lc app=leetcode.cn id=133 lang=cpp
 *
 * [133] 克隆图
 */

#include <bits/stdc++.h>
using namespace std;
// @lc code=start

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) return nullptr;
        Node* res = new Node(node->val);
        unordered_map<Node*, Node*> m;
        m[node] = res;
        queue<Node*> q;
        q.push(node);
        while (!q.empty()) {
            Node* cur = q.front();
            q.pop();
            for (auto& neighbor : cur->neighbors) {
                if (!m.count(neighbor)) {
                    m[neighbor] = new Node(neighbor->val);
                    q.push(neighbor);
                }
                m[cur]->neighbors.push_back(m[neighbor]);
            }
        }
        return res;
    }
};
// @lc code=end

